aten/src/ATen/native/cpu/UnaryOpsKernel.cpp

```diff
@@ -661,6 +661,37 @@ static void rsqrt_kernel(TensorIterator& iter) {
   }                                                                           \
   REGISTER_DISPATCH(op##_stub, &op##_kernel)
 
+  #define IMPLEMENT_COMPLEX_STRUCTURED_KERNEL(dispatchtypes, op)                             \
```

Comment:
cc @jamesr66a I may be going crazy here, but is there a difference between `TensorIterator` and `TensorIteratorBase`? It seems that I'm getting different results with each. I've just changed the former to the latter and additionally added `_STRUCTURED` to the macro.

Reply:
Getting these failures:

```
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_bfloat16 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=0.01, found 4084 ele...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_bool - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 2856 element(...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_complex128 - AssertionError: False is not true : Tensors failed to compare as equal!Real parts failed to compare as equal! With r...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_complex64 - AssertionError: False is not true : Tensors failed to compare as equal!Real parts failed to compare as equal! With rt...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_float32 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 4096 eleme...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_float64 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 4096 ele...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int16 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3930 element...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int32 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3882 element...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int64 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3882 element...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int8 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3882 element(...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_uint8 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3737 element...

```

Reply:
I'm probably not the best person to ask. @ezyang @colesbury @cpuhrsch maybe?

Reply:
I'm going to take a look at this

Reply:
It looks like adding in a call to `TensorIteratorBase::cast_outputs()` should fix those tests. It was getting called by `unary_op_impl_float_out` originally, but isn't there after the changes in this PR.

The tests that were failing all dealt with `out` tensors that had a different dtype than the inputs. Currently when TensorIterator encounters that, it looks like it creates a temp output tensor and expects you to call `cast_outputs()` to copy the temp into the actual output tensor.

The reason you saw the failure here but we didn't see an issue porting `add_out` to structured is because binary ops all currently call `cpu_kernel()` in `Loops.h`, which handles calling `cast_outputs()`, while the unary ops in this file call `cast_outputs()` directly.

It would be nice to get this to be done more automatically by TensorIterator: it looks like Ed left a comment [here](https://github.com/pytorch/pytorch/blob/a0652c8f08f5257735f87106363258d1c14a2b52/aten/src/ATen/TensorIterator.cpp#L411) pretty much saying that, although it might take some more refactoring of TensorIterator (probably not worth doing right now). (cc @bhosmer)

I can see two reasonable ways to add it in and finish the port:
- Add a call to `iter.cast_outputs()` at the end of this `IMPLEMENT_COMPLEX_STRUCTURED_KERNEL` macro. I tried that locally and it looks like tests pass.
- Update the macro to call `cpu_kernel` / `cpu_kernel_vec` from `Loops.h`, which already has logic to call cast_outputs(). That would probably be better to cut down on code duplication, but I'm not sure how much of a drop-in code replacement that would be (i.e. if it'd break some tests).

Reply:
and I *think* adding in the duplicate macro is reasonable to deal with any differences between them, until all of the unary ops are ported (especially given that you'll need to update the structured one)

Reply:
Thanks for the detailed explanation, @bdhirsh! Using `cpu_kernel_vec` fixed the test failures locally, let's see if it causes any more failures. 🤞🏼

Reply:
Resolving this, since some CPU tests have fully passed.